### ISSUE TITLE
Somthing not right when source and sinks defined as same time

### ISSUE THREAD

**Issue by jgostick** [PMEAL/OpenPNM#1162]
This following code repoduces this bug:

```
import openpnm as op
import scipy as sp
import matplotlib.pyplot as plt
pn = op.network.Cubic(shape=[40, 40], spacing=1e-4)
geo = op.geometry.StickAndBall(network=pn, pores=pn.Ps, throats=pn.Ts)
gas = op.phases.Air(network=pn)
phys = op.physics.Standard(network=pn, phase=gas, geometry=geo)
gas['pore.concentration'] = 0
phys['pore.sinkA'] = -1e10
phys['pore.sinkb'] = 1
phys.add_model(propname='pore.sink', model=op.models.physics.generic_source_term.power_law,
               A1='pore.sinkA', A2='pore.sinkb', X='pore.concentration')
phys['pore.srcA'] = 1e-10
phys['pore.srcb'] = 1
phys.add_model(propname='pore.source', model=op.models.physics.generic_source_term.power_law,
               A1='pore.srcA', A2='pore.srcb', X='pore.concentration')
rx = op.algorithms.FickianDiffusion(network=pn)
rx.setup(phase=gas)
rx.set_source(propname='pore.source', pores=[410, 610, 810])  # Comment this line on and off to see problem
rx.set_source(propname='pore.sink', pores=[430, 630, 830])
rx.set_value_BC(values=1, pores=pn.pores('front'))
rx.run()
im = sp.reshape(rx['pore.concentration'], [40, 40]).T
fig = plt.figure(figsize=[10, 10])
plt.imshow(im)
```

**Comment by ma-sadeghi** [comment]
@jgostick If you change the magnitude of the source and sink terms, the concentration profile seems to be reasonable (I tried with 1e-11 for both source and sink). Something goes wrong when the source term (and only the source term, since the sink term seems to be very robust at various orders of magnitudes) gets bigger than some value, which results in unrealistic concentrations (negative in this case).

To sum up, I guess we need to investigate how **source** term is being applied. It's possible that this is a classical debate/issue in CFD, and might have already been addressed.

**Comment by jgostick** [comment]
Yes it seems to be the source term that is the problem.  I'll see if @magnaou has any ideas.

**Comment by magnaou** [comment]
I checked the numerical models and the code. It looks like everything is done properly.
The example Jeff posted above (which is a linear source/sink term) doesn't converge with certain linear solvers such as "pyamg". I doubt the solution with negative concentrations we are getting is a converged one. Is this due to the fact that the problem (network, geometry, physical properties, BCs, etc) is not physical/realistic?
I'm taking a look to methods for bounding the concentration...
